src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -478,8 +478,9 @@ public async Task ConnectAsync(CancellationToken cancellation = default(Cancella
         /// <param name="previous">Previous network state of the peer.</param>
         private async Task OnStateChangedAsync(NetworkPeerState previous)
         {
-            bool insideCallback = this.onDisconnectedAsyncContext.Value == null;
-            if (!insideCallback)
+            // Creates a context that can be used for postponing/flagging disconnect.
+            bool iCreatedContext = this.onDisconnectedAsyncContext.Value == null;
```

Comment:
Is this variable name correct? if `this.onDisconnectedAsyncContext.Value == null` then it means I did not create the context (yet)

Reply:
The purpose of the AyncLocal here is to track whether the Task that started to processes the message and entered the behaviours for processing is called form a behaviour itself (if I remember correctly) if calling disconnect this can change the sate and call back in to this method.

Reply:
> Is this variable name correct? if this.onDisconnectedAsyncContext.Value == null then it means I did not create the context (yet)

It's equivalent to this code where the naming makes sense:

```
            bool iCreatedContext = false;

            if (this.onDisconnectedAsyncContext.Value == null)
            {
                this.onDisconnectedAsyncContext.Value = new DisconnectedExecutionAsyncContext();
                iCreatedContext = true;
            }
```

Reply:
> The purpose of the AyncLocal here is to track whether the Task that started to processes the message and entered the behaviours for processing is called form a behaviour itself (if I remember correctly) if calling disconnect this can change the sate and call back in to this method.

The code does not make any assumptions as to what is happening externally and neither should it have to.